### PR TITLE
IFC-618 fix diff relationship bug when conflicting changes with main changes after branch changes

### PR DESCRIPTION
**Pull request by ajtmccarty**
IFC-618

fix a bug in diff calculation that affected relationships of cardinality=one when a conflicting change was made on the base branch AFTER the change on the diff branch

### CODE REVIEW

File: backend/infrahub/core/diff/enricher/cardinality_one.py [hunk 1]
```diff
@@ -33,27 +33,15 @@ def __init__(self, db: InfrahubDatabase):
...
 
-    def is_cardinality_one(self, node_kind: str, relationship_name: str, diff_branch_name: str) -> bool:
-        if node_kind not in self._node_schema_map:
-            self._node_schema_map[node_kind] = self.db.schema.get(
-                name=node_kind, branch=diff_branch_name, duplicate=False
-            )
-        node_schema = self._node_schema_map[node_kind]
-        relationship_schema = node_schema.get_relationship(name=relationship_name)
-        return relationship_schema.cardinality == RelationshipCardinality.ONE
-
```
Comment (ajtmccarty) [inline backend/infrahub/core/diff/enricher/cardinality_one.py:56]: this calculation is no longer required b/c we have the cardinality in the diff element already

File: backend/infrahub/core/diff/query_parser.py [hunk 2]
```diff
@@ -551,6 +554,14 @@ def _apply_relationship_previous_values(
...
                 }
                 base_diff_property_by_type[DatabaseEdgeType.IS_RELATED] = None
+                latest_diff_property_times_by_type: dict[DatabaseEdgeType, Timestamp] = {}
+                for diff_property in property_set:
+                    if (
+                        diff_property.property_type not in latest_diff_property_times_by_type
+                        or diff_property.changed_at > latest_diff_property_times_by_type[diff_property.property_type]
+                    ):
+                        latest_diff_property_times_by_type[diff_property.property_type] = diff_property.changed_at
+
```
Comment (ajtmccarty) [inline backend/infrahub/core/diff/query_parser.py:561]: this new logic updates how we apply the changes on the base branch to the diff branch when setting previous values in the diff. the problem was that if the changes on `main` occurred after the changes on `branch`, then we could end up treating later changes as though they preceded the changes on the branch.
this new dict and conditional to check on it fix that by ensuring that changes on main that occurred after the changes on the branch are ignored

File: backend/tests/integration/diff/test_diff_incremental_addition.py
Comment (ajtmccarty) [inline backend/tests/integration/diff/test_diff_incremental_addition.py:1]: the updates in here are to ensure that our diff calculations are the same if they are incremental or all-at-once for a given set of changes

File: backend/tests/unit/core/diff/test_diff_query_parser.py [hunk 3]
```diff
@@ -1013,6 +1013,276 @@ async def test_many_relationship_property_update(
     assert source_rel.new_value == person_jane_main.get_id()
 
 
+async def test_cardinality_one_peer_conflicting_updates(
```
Comment (ajtmccarty) [inline backend/tests/unit/core/diff/test_diff_query_parser.py:1016]: new test for the actual bug

File: backend/tests/integration/diff/test_diff_incremental_addition.py [hunk 4]
```diff
@@ -84,46 +87,126 @@ async def diff_coordinator(self, db: InfrahubDatabase, diff_branch: Branch) -> D
...
         default_branch: Branch,
         diff_branch: Branch,
-        diff_coordinator: DiffCoordinator,
+        data_02_previous_owner_on_branch,
+    ) -> None:
+        delorean = initial_dataset["delorean"]
+        biff = initial_dataset["biff"]
+
+        delorean_main = await NodeManager.get_one(db=db, branch=default_branch, id=delorean.get_id())
+        await delorean_main.previous_owner.update(db=db, data=[{"id": biff.get_id(), "_relation__is_protected": True}])
```
Comment (ogenstad) [inline backend/tests/integration/diff/test_diff_incremental_addition.py:135]: Note that Biff never owned a Delorean he just stole one.

**Review by ogenstad — APPROVED** [review]

**Pull request merged**